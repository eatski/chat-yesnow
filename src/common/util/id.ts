import { generate } from "randomstring"

export const generateId = () => {
    return generate(15);
}