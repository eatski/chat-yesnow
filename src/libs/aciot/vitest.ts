import { AxiosInstance } from "axios";
import { AciotCore, CacheMode } from "./core";
import { afterAll, beforeAll } from "vitest";
import path from "path";
import filenamify from 'filenamify';

export const applyTestHooks = (
	axios: AxiosInstance,
	config: {
		mode: CacheMode;
	},
) => {
	let aciot: AciotCore | null;
	beforeAll(({ name, file }) => {
		if (!file?.name) {
			throw new Error("suite.file.name is not defined");
		}
		const parsedPath = path.parse(file.name);
		const filename = filenamify(name, { replacement: "_" });
		aciot = new AciotCore({
			cacheBasePath: path.resolve(
				process.cwd(),
				parsedPath.dir,
				"__data__",
				parsedPath.name + parsedPath.ext,
				filename
			),
		});
		aciot.applyInterceptors(axios, config.mode);
	});
	afterAll(async ({ name, tasks }) => {
		if (tasks.every((t) => t.result?.state === "pass")) {
			console.info(name, "Clearing unused cache...");
			await aciot?.clearUnusedCache();
		}
	});
};
