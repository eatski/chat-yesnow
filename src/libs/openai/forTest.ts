import { Configuration, OpenAIApi } from "openai";
import Axios,{ AxiosRequestConfig} from "axios";
import Cache from "file-system-cache";
import { resolve } from "path";

const cache = Cache({
    basePath: resolve(process.cwd(), ".resultSnapshots"),
});

class MyError extends Error {
    constructor(public readonly key: string) {
        super("MyError");
    }
}

const configToKey = (config: AxiosRequestConfig): string => {
    return JSON.stringify({
        url: config.url,
        body: config.data,
    });
}

const axios = Axios.create();
if(process.env.CI){
    axios.interceptors.request.use((config) => {
        const key = configToKey(config)
        return Promise.reject(new MyError(key));
    })
    axios.interceptors.response.use(() => {},async (rejected) => {
        if(rejected instanceof MyError){
            return Promise.resolve({
                data: await cache.get(rejected.key)
            })
        }
        return Promise.reject(rejected)
    })
} else {
    axios.interceptors.response.use((res) => {
        const key = configToKey(res.config);
        res.data.id = "test";
        res.data.created = new Date("2021-01-01");
        cache.set(key, res.data);
        return res
    })
}


export const openaiForTest = new OpenAIApi(
    new Configuration({
        apiKey: process.env.OPENAI_API_KEY,
    }),
    undefined,
    axios
)