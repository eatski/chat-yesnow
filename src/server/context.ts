import { authConfig } from "@/pages/api/auth/[...nextauth]";
import { TRPCError } from "@trpc/server";
import { CreateNextContextOptions } from "@trpc/server/adapters/next";
import { getServerSession } from "next-auth/next";
import { setTimeout } from "timers/promises";
import { verifyRecaptcha } from "./services/recaptcha";
import {
	AB_TESTING_COOKIE_NAME,
	AB_TESTING_VARIANTS,
	getAorBRandom,
	validateABTestingVariant,
} from "@/common/abtesting";
import { get } from "@vercel/edge-config";

export const createContext = async (context: CreateNextContextOptions) => {
	return {
		setupABTestingVariant: async () => {
			const abTestRate = await get("abTestRate");
			if (
				!abTestRate ||
				!(typeof abTestRate === "number") ||
				isNaN(abTestRate) ||
				abTestRate < 0 ||
				abTestRate > 1
			) {
				return AB_TESTING_VARIANTS.ONLY_SONNET;
			}
			const cookieValue = context.req.cookies[AB_TESTING_COOKIE_NAME];
			const validated = cookieValue && validateABTestingVariant(cookieValue);
			if (validated) {
				return validated;
			}
			const random = getAorBRandom(abTestRate);
			context.res.setHeader(
				"Set-Cookie",
				`${AB_TESTING_COOKIE_NAME}=${random}; Path=/; HttpOnly; Secure; SameSite=Strict`,
			);
			return random;
		},
		getUserOptional: async () => {
			const session = await getServerSession(
				context.req,
				context.res,
				authConfig,
			);
			return session?.user?.email && session.custom?.userId
				? {
						id: session.custom.userId,
						email: session.user.email,
				  }
				: null;
		},
		getUser: async () => {
			const session = await getServerSession(
				context.req,
				context.res,
				authConfig,
			);
			if (!session?.user?.email || !session.custom?.userId) {
				throw new TRPCError({
					code: "UNAUTHORIZED",
				});
			}
			return {
				id: session.custom.userId,
				email: session.user.email,
			};
		},
		doRevalidate(url: string) {
			const retryable = (count: number, error?: any): Promise<void> => {
				if (count) {
					return setTimeout(3000).then(() =>
						context.res.revalidate(url).catch((e) => {
							return retryable(count - 1, e);
						}),
					);
				}
				console.error(error);
				return Promise.reject(
					new Error("revalidation failed", {
						cause: error,
					}),
				);
			};
			return retryable(10);
		},
		verifyRecaptcha(token: string) {
			return verifyRecaptcha(token).catch((e) => {
				throw new TRPCError({
					code: "BAD_REQUEST",
					cause: e,
				});
			});
		},
		isThankYouUser: () =>
			context.req.cookies.thankyou === process.env.THANKYOU_CODE,
	};
};
