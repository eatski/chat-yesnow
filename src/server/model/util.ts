export type Public<T, F extends keyof T> = {
	[K in F]: DateToNumber<T[K]>;
} & {
	[K in Exclude<keyof T, F>]?: never;
};

export type Override<
	T extends Record<keyof U, unknown>,
	U extends Partial<Record<keyof T, unknown>>,
> = Omit<T, keyof U> & U;

type DateToNumber<T> = T extends Date ? number : T;
