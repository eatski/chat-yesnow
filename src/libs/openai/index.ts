import { OpenAI } from "openai";
import {
	ChatCompletionCreateParamsNonStreaming,
	EmbeddingCreateParams,
} from "openai/resources";
const openai = new OpenAI({
	// fetchを直接使うとmswの上書きが間に合わないので、fetchのwrapperにする。
	fetch: (...args) => fetch(...args),
	timeout: 20000,
	apiKey: process.env.OPENAI_API_KEY || "dummy",
});

export const createOpenAICompletion = async (
	config: ChatCompletionCreateParamsNonStreaming,
) => {
	return openai.chat.completions.create(config);
};

export const createOpenAIEmbedding = async (input: EmbeddingCreateParams) => {
	return openai.embeddings.create(input);
};
