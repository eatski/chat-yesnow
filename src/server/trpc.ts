import { authConfig } from '@/pages/api/auth/[...nextauth]';
import { initTRPC, TRPCError } from '@trpc/server';
import { CreateNextContextOptions } from '@trpc/server/adapters/next';
import { getServerSession } from 'next-auth/next';
import { setTimeout } from 'timers/promises';

export const createContext = async (context: CreateNextContextOptions) => {
    const session = await getServerSession(context.req, context.res, authConfig);
    if(!session || !session.user?.email){
        throw new TRPCError({
            code: 'UNAUTHORIZED',
        })
    }
    return {
      user: {
        email: session.user.email,
      },
      doRevalidate(url: string){
        const retryable = (count: number,error?: any): Promise<void> => {
          if(count){
            return setTimeout(3000).then(() => context.res.revalidate(url).catch((e) => {
              return retryable(count - 1,e)
            }))
          }
          console.error(error);
          return Promise.reject( new Error("revalidation failed",{
            cause: error
          }))
        }
        return retryable(10)
      },
    };
};

/**
 * Initialization of tRPC backend
 * Should be done only once per backend!
 */
const t = initTRPC.context<typeof createContext>().create();

/**
 * Export reusable router and procedure helpers
 * that can be used throughout the router
 */
export const router = t.router;
export const procedure = t.procedure;