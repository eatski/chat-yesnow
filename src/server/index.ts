import { hello } from './procedures/hello';
import { post } from './procedures/post';
import { put } from './procedures/put';
import { question } from './procedures/question';
import { truth } from './procedures/truth';
import { router } from './trpc';

export const appRouter = router({
  hello,
  question,
  truth,
  post,
  put
});

// Export type router type signature,
// NOT the router itself.
export type AppRouter = typeof appRouter;