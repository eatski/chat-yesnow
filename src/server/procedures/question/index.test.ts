import { describe, test, expect, beforeAll } from "vitest";
import { appRouter } from "@/server";
import { setupOpenaiForTest } from "@/libs/openai/forTest";
import { prepareStoryFromYaml } from "@/test/prepareStory";
import { resolve } from "path";
import { generateId } from "@/common/util/id";
const never = () => {
	throw new Error("Never");
};

const TEST_ID = generateId();
const TEST_ID_PRIVATE = generateId();

describe("trpc/question", () => {
	const testYamlPath = resolve(process.cwd(), "fixtures", "test.yaml");
	const TEST_USER1 = {
		id: generateId(),
		email: "yesonor@example.com",
	};

	const TEST_USER2 = {
		id: generateId(),
		email: "not-yesonor@example.com",
	};
	beforeAll(async () => {
		const cleanup1 = await prepareStoryFromYaml(testYamlPath, {
			storyId: TEST_ID,
			authorId: TEST_USER1.id,
			published: true,
		});
		const cleanup2 = await prepareStoryFromYaml(testYamlPath, {
			storyId: TEST_ID_PRIVATE,
			authorId: TEST_USER2.id,
			published: false,
		});
		return async () => {
			await cleanup1();
			await cleanup2();
		};
	});
	const openai = setupOpenaiForTest();
	const testee = appRouter.createCaller({
		getUserOptional: async () => null,
		getUser: never,
		doRevalidate: never,
		verifyRecaptcha: () => Promise.resolve(),
		openai,
	});
	describe("質問した内容に対して、結果が返る", () => {
		test.concurrent.each([
			"山田さんは犯罪者ですか？",
			"人が死んでますか？",
			"近くに他の人がいますか？",
		])("真相に対して正しい質問をするとTrueが返る", async (text) => {
			const result = await testee.question({
				storyId: TEST_ID,
				text,
				recaptchaToken: "anytoken",
			});
			expect(result.answer).toEqual("True");
		});
		test.concurrent.each([
			"山田さんは誰かに襲われてますか？",
			"性的な興味で女性用トイレに入りましたか？",
			"山田は死んでいますか？",
		])("真相に対して正しくない質問をするとFalseが返る", async (text) => {
			const result = await testee.question({
				storyId: TEST_ID,
				text,
				recaptchaToken: "anytoken",
			});
			expect(result.answer).toEqual("False");
		});
		test.concurrent.each([
			"山田には恋人がいますか？",
			"山田さんはパンよりライス派？",
		])("真相では言及されていない質問をするとUnknownが返る", async (text) => {
			const result = await testee.question({
				storyId: TEST_ID,
				text,
				recaptchaToken: "anytoken",
			});
			expect(result.answer).toEqual("Unknown");
		});
	});
	describe("customMessage", () => {
		test("customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る", async () => {
			const result = await testee.question({
				storyId: TEST_ID,
				text: "人を殺しましたか？",
				recaptchaToken: "anytoken",
			});
			expect(result.hitQuestionExample?.customMessage).toMatchInlineSnapshot(
				"undefined",
			);
		});
		test("customMessageを持つquestionExamlpeに近しくない質問をすると、customMessageが返らない", async () => {
			const result = await testee.question({
				storyId: TEST_ID,
				text: "山田さんは犯罪者ですか？",
				recaptchaToken: "anytoken",
			});
			expect(result.answer).toEqual("True"); //FIXME: Falseのはず
			expect(result.hitQuestionExample?.customMessage).not.toBeDefined();
		});
	});
	describe("storyの参照権限", () => {
		test.each([TEST_USER1, TEST_USER2, null])(
			"publicなstoryに対して質問すると回答が返ってくる",
			async (user) => {
				const testee = appRouter.createCaller({
					getUserOptional: async () => user,
					getUser: never,
					doRevalidate: never,
					verifyRecaptcha: () => Promise.resolve(),
					openai,
				});
				const text = "人を殺しましたか？";
				const result = await testee.question({
					storyId: TEST_ID,
					text,
					recaptchaToken: "anytoken",
				});
				expect(result.answer).toEqual("True");
			},
		);
		test.each([TEST_USER1, null])(
			"privateなstoryに対して質問すると回答が返ってこない",
			async (user) => {
				const testee = appRouter.createCaller({
					getUserOptional: async () => user,
					getUser: never,
					doRevalidate: never,
					verifyRecaptcha: () => Promise.resolve(),
					openai,
				});
				const text = "人を殺しましたか？";
				expect(
					testee.question({
						storyId: TEST_ID_PRIVATE,
						text,
						recaptchaToken: "anytoken",
					}),
				).rejects.toMatchInlineSnapshot("[TRPCError: NOT_FOUND]");
			},
		);
		test("自分の作成したstoryの場合、privateなstoryに対して質問しても回答が返ってくる", async () => {
			const testee = appRouter.createCaller({
				getUserOptional: async () => TEST_USER2,
				getUser: never,
				doRevalidate: never,
				verifyRecaptcha: () => Promise.resolve(),
				openai,
			});
			const text = "人を殺しましたか？";
			const result = await testee.question({
				storyId: TEST_ID_PRIVATE,
				text,
				recaptchaToken: "anytoken",
			});
			expect(result.answer).toEqual("True");
		});
	});
});
