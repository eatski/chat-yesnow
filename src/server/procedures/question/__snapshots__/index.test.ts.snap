// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 4`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 5`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 6`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question ONLY_SONNET > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question ONLY_SONNET > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question ONLY_SONNET > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question WITH_HAIKU > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question WITH_HAIKU > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question WITH_HAIKU > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=%s > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=%s > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=%s > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=A > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=A > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=A > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=B > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 1`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=B > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 2`] = `"いい質問ですね！山田さんは人を殺しています。"`;

exports[`trpc/question abtesting=B > customMessage > customMessageを持つquestionExamlpeに近しい質問をすると、customMessageが返る 3`] = `"いい質問ですね！山田さんは人を殺しています。"`;
